servicetalk-loadbalancer-experimental/src/main/java/io/servicetalk/loadbalancer/DefaultLoadBalancer.java

```diff
@@ -287,16 +278,27 @@ public void onSubscribe(final Subscription s) {
         @Override
         public void onNext(@Nullable final Collection<? extends ServiceDiscovererEvent<ResolvedAddress>> events) {
             if (events == null || events.isEmpty()) {
-                LOGGER.debug("{}: unexpectedly received null or empty list instead of events.",
-                        DefaultLoadBalancer.this);
+                LOGGER.debug("{}: unexpectedly received null or empty collection instead of events: {}",
+                        DefaultLoadBalancer.this, events);
                 return;
             }
             sequentialExecutor.execute(() -> sequentialOnNext(events));
         }
 
         private void sequentialOnNext(Collection<? extends ServiceDiscovererEvent<ResolvedAddress>> events) {
-            if (isClosed || events.isEmpty()) {
-                // nothing to do if the load balancer is closed or there are no events.
+            assert !events.isEmpty();
+            if (isClosed) {
+                // nothing to do if the load balancer is closed.
+                return;
+            }
+
+            // According to Reactive Streams Rule 1.8
+            // (https://github.com/reactive-streams/reactive-streams-jvm?tab=readme-ov-file#1.8) new events will
+            // stop eventually but not guaranteed to stop immediately after cancellation or could race with cancel.
+            // Therefore, we should check that this is the current Subscriber before processing new events.
+            if (healthCheckConfig != null && currentSubscriber != this) {
```

Comment:
This is my only real concern with the PR.
```suggestion
            if (currentSubscriber != this) {
```
It seems to me that it shouldn't matter for what reason we switched our subscription: this one is simply no longer active so we must ignore updates or risk clobbering updates from the active subscription.
Same in RoundRobinLoadBalancer.